scrapy/utils/request.py

```diff
@@ -68,6 +70,67 @@ def request_fingerprint(request, include_headers=None, keep_fragments=False):
     return cache[cache_key]
 
 
+_fingerprint_cache = weakref.WeakKeyDictionary()
+
+
+def fingerprint(request, *, include_headers=None, keep_fragments=False) -> bytes:
+    """
+    Return the request fingerprint.
+
+    The request fingerprint is a hash that uniquely identifies the resource the
+    request points to. For example, take the following two urls:
+
+    http://www.example.com/query?id=111&cat=222
+    http://www.example.com/query?cat=222&id=111
+
+    Even though those are two different URLs both point to the same resource
+    and are equivalent (i.e. they should return the same response).
+
+    Another example are cookies used to store session ids. Suppose the
+    following page is only accessible to authenticated users:
+
+    http://www.example.com/members/offers.html
+
+    Lots of sites use a cookie to store the session id, which adds a random
+    component to the HTTP Request and thus should be ignored when calculating
+    the fingerprint.
+
+    For this reason, request headers are ignored by default when calculating
+    the fingeprint. If you want to include specific headers use the
+    include_headers argument, which is a list of Request headers to include.
+
+    Also, servers usually ignore fragments in urls when handling requests,
+    so they are also ignored by default when calculating the fingerprint.
+    If you want to include them, set the keep_fragments argument to True
+    (for instance when handling requests with a headless browser).
+    """
+    if include_headers:
+        include_headers = tuple(to_bytes(h.lower()) for h in sorted(include_headers))
+    cache = _fingerprint_cache.setdefault(request, {})
+    cache_key = (include_headers, keep_fragments)
+    if cache_key not in cache:
+        fp = hashlib.sha1()
+        fp.update(to_bytes(request.method))
+        fp.update(to_bytes(canonicalize_url(request.url, keep_fragments=keep_fragments)))
+        fp.update(request.body or b'')
+        if include_headers:
+            for hdr in include_headers:
+                if hdr in request.headers:
+                    fp.update(hdr)
+                    for v in request.headers.getlist(hdr):
+                        fp.update(v)
```

Comment:
Sorry, I'm getting more an more unrelated comments on this PR :) This is a nitpick, but it looks like body & headers & values are not separated where fp is being built. So e.g. a request with header with "foo" and "bar" values would get the same fingerprint as a request with "fo" and "obar" values. The same applies for body, etc. URL is joined with the body, for example, so a POST request to http://example.com with empty body may get the same fingerprint as a POST request to http://example.co with "m" body.

Maybe it was not invisible in past, because such cases are rare, and because include_headers was a bit tricky to use. Now we're documenting how to use it, so the issue may become a bit more important. 

Changing the fingerprint for a request should be considered a backwards incompatible change - e.g. disk cache will be effectively emptied after Scrapy upgrade. Here we're introducing a new function, so there can be an argument that we should make the change now, to not change the new function in future. But I think this can be done separately - a real breaking change is not the output of fingerprint function per se, but fingerprints requests get by default, so we can't avoid making this non-breaking. Opened a placeholder https://github.com/scrapy/scrapy/issues/4762 for that.

Reply:
The main problem is that the new function is used by the (new) default request fingerprinter.

To avoid users using this function for their custom request fingerprinters, we could:
#. Create a private copy of the function, for the default fingerprinter to use to keep backward compatibility.
#. Fix the function.

That’s something that I believe would make sense to do as part of this pull request. Thoughts?

Then there’s the issue of moving users from the bugged, default fingerprinter logic, to one that is not bugged. I think that whether we make that change as part of this pull request or we handle it as part of #4762 depends on how we want to approach it:

- We could modify this pull request to introduce 2 different fingerprinter classes, one with the deprecated behavior (e.g. `DeprecatedRequestFingerprinter`, `Unsafe…`, `Bogus…`, `Old…`, `Pre2Dot4…`), and one with the fixed behavior (i.e. `RequestFingerprinter`), and use the deprecated class by default, and have it log a warning asking users to switch to the new one. For projects generated with `startproject`, we could set that automatically.
- We could do something similar, but using a setting which the default fingerprinter reads. In that case, we _can_ handle it separately as a fix for #4762, although I would not mind fixing it here.